code/game/machinery/clonepod.dm

```diff
@@ -28,368 +55,491 @@ GLOBAL_LIST_INIT(cloner_biomass_items, list(\
-		to_chat(H, "<span class='notice'>[message]</span>")
-	else if(grab_ghost_when == CLONER_MATURE_CLONE)
-		to_chat(clonemind.current, "<span class='notice'>Your body is beginning to regenerate in a cloning pod. You will become conscious when it is complete.</span>")
-		// Set up a soul link with the dead body to catch a revival
-		RegisterSignal(clonemind.current, COMSIG_LIVING_REVIVE, PROC_REF(occupant_got_revived))
-		RegisterSignal(clonemind, COMSIG_MIND_TRANSER_TO, PROC_REF(occupant_got_revived))
-
-	SSblackbox.record_feedback("tally", "players_revived", 1, "cloned")
-	update_icon()
+//Process
+/obj/machinery/clonepod/process()
 
-	H.suiciding = FALSE
-	attempting = FALSE
-	return 1
+	//Basically just isolate_reagent() with extra functionality.
+	for(var/datum/reagent/R as anything in reagents.reagent_list)
+		if(!(R.id in VALID_REAGENTS))
+			reagents.del_reagent(R.id)
+			reagents.update_total()
+			atom_say("Purged contaminant from chemical storage.")
 
-//Grow clones to maturity then kick them out. FREELOADERS
-/obj/machinery/clonepod/process()
+	//Take in biomass. Mostly copied from the old cloning code
 	var/show_message = FALSE
 	for(var/obj/item/item in range(1, src))
-		if(is_type_in_list(item, GLOB.cloner_biomass_items))
+		if(is_type_in_list(item, VALID_BIOMASSABLES) && (biomass + MEAT_BIOMASS_VALUE <= biomass_storage_capacity))
 			qdel(item)
-			biomass += BIOMASS_BASE_AMOUNT
+			biomass += MEAT_BIOMASS_VALUE
 			show_message = TRUE
 	if(show_message)
-		visible_message("[src] sucks in and processes the nearby biomass.")
+		visible_message("[src] sucks in nearby biomass.")
+
+	//If we're cloning someone, we haven't generated a list of limbs to grow, and we're before any possibility of not having any limbs left to grow.
+	if(currently_cloning && !length(limbs_to_grow) && clone_progress < 20)
+		limb_loop:
+			for(var/limb in desired_data.limbs)
+				if(desired_data.limbs[limb][4])
+					continue //We're not growing this limb, since in the desired state it's missing.
+
+				var/obj/item/organ/external/limb_typepath = patient_data.genetic_info.species.has_limbs[limb]["path"]
+				if(initial(limb_typepath.vital)) //I hate everything about this check, but it sees if the current organ is vital..
+					continue //and continues if it is, since the proc that creates the clone mob will make these all at once.
+
+				for(var/organ in desired_data.organs)
+					var/obj/item/organ/external/organ_typepath = patient_data.genetic_info.species.has_organ[organ]
+					if(!initial(organ_typepath.vital)) //I hate this check too. We loop through all the organs the cloned species should have.
+						continue //If it's not a vital organ, continue looping through the organs.
+					if(initial(organ_typepath.parent_organ) == limb)
+						continue limb_loop //If it's a vital organ, and belongs to the current limb, we don't want this limb.
+
+				limbs_to_grow += limb //It's not supposed to be missing and it's not vital - so we'll be growing it.
+
```

Comment:
```suggestion
		for(var/limb in desired_data.limbs)
			if(desired_data.limbs[limb][4])
				continue //We're not growing this limb, since in the desired state it's missing.

			var/obj/item/organ/external/limb_typepath = patient_data.genetic_info.species.has_limbs[limb]["path"]
			if(initial(limb_typepath.vital)) //I hate everything about this check, but it sees if the current organ is vital..
				continue //and continues if it is, since the proc that creates the clone mob will make these all at once.
			var/parent_organ_is_limb = FALSE
			for(var/organ in desired_data.organs)
				var/obj/item/organ/external/organ_typepath = patient_data.genetic_info.species.has_organ[organ]
				if(!initial(organ_typepath.vital)) //I hate this check too. We loop through all the organs the cloned species should have.
					continue //If it's not a vital organ, continue looping through the organs.
				if(initial(organ_typepath.parent_organ) == limb)
					parent_organ_is_limb = TRUE //If it's a vital organ, and belongs to the current limb, we don't want this limb.
					break
			if(parent_organ_is_limb)
				continue
			limbs_to_grow += limb //It's not supposed to be missing and it's not vital - so we'll be growing it.

```
code that doesn't use loop labeling syntax